### ISSUE TITLE
508-defect-1 ❗ Launchblocker [SCREEN READER, COGNITION]: The auto-suggest search field MUST be accessible to screen reader users

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#16965]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

<hr/>

## Point of Contact

**VFS Point of Contact:**  Josh

## User Story or Problem Statement

As a user with limited or no vision who uses a screen reader, I want my screen reader to convey to me what others can see on a page, so I can understand the structure, information and relationships on a page. For example, I want my screen reader to announce:

- an autosuggest exists
- how many items are being suggested
- when it has been activated or deactivated (expanded/closed)

As a keyboard user, I want to be able to navigate to the auto suggested items using my arrow keys so that I can access them without having to tab over the search button and throughout the list itself.

## Details

There's several problems to be solved with this auto-suggest component including:
- It has multiple `tab` stops which does not meet [WAI-ARIA keyboard specs](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_generalnav).
- It is not accessibly labeled or identified as a combobox or search.
- It does not announce to screen reader users that suggestions exist.
- It does not provide instructions on how to navigate suggestions.

## Acceptance Criteria

- [x] The solution follows [ARIA 1.2 Combobox guidance](https://www.w3.org/TR/wai-aria-practices-1.2/examples/combobox/combobox-autocomplete-list.html).
- [x] The user can navigate to the suggestions using the `up` and `down` arrow keys and escape using the `esc` key.
- [x] The screen reader politely announces when auto suggestions exist and how many there are.
- [x] Zero axe plugin violations when the typeahead is present, and when the typeahead is hidden
  - `aria-controls` throws an error if it references an ID that's not present in the DOM. You may need to add this attribute when the list exists, and remove it when the list is hidden.

## Environment

* Operating System: Mac
* Browser: Safari
* Screenreading device: Voice Over
* Server destination: staging

## Steps to Recreate

1. Enter `<https://staging.va.gov/>` in browser.
2. Start screenreading device listed in Environment.
3. Navigate to the search navigational item.
4. Open the search with `space`.
5. Type "Vet" into the search field.
6. Confirm the screen reader does not announce the auto-suggestions.
7. Confirm suggestions are not keyboard accessible to `up` and `down` .

## Previous Work - Datalist and Downshift
The full history of the type ahead / auto-suggest / auto-complete component can be found in the [[SD] Type Ahead POC epic](https://github.com/department-of-veterans-affairs/va.gov-team/issues/8896).

Current implementations of the type ahead can be found in John Hashimoto's documentation [On Site Search - Type-Ahead Design + UX issue](https://github.com/department-of-veterans-affairs/va.gov-team/issues/10489).

To review consistency across various implementations, Liz Lantz posted this into the [DSVA #design slack channel](https://app.slack.com/client/T03FECE8V/C0NGDDXME/thread/C0NGDDXME-1592834606.458500) where @jenstrickland  [suggested the use of a data-list in a code pen she authored](https://codepen.io/jenstrickland/pen/XWdwjEJ).

The datalist codepen was then [referenced by Nick Sullivan in this DSVA slack thread](https://dsva.slack.com/archives/C011K1VSTC3/p1592853360008800) and implemented per the accessibility recommendation. [Screenshots and description of the implementation were written in a comment by Nick.](https://github.com/department-of-veterans-affairs/va.gov-team/issues/10489#issuecomment-672053810)

Crystabel Reiter followed up on this suggestion by creating the feature request ticket: [Standardize type ahead search pattern](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/308). Here she notes that two options are being considered based on user testing and accessibility testing:
- A datalist component, which requires more research on making option text larger as users zoom their browser along with styling the option elements.
- Using a current version of Downshift which is still under active development and includes new React hook-based options. This would be a significant time investment, but could be more extensible in the long term. @1Copenut has built a custom typeahead using Downshift in the past and contributed to this component as a reviewer. [He recommended abiding by the WAI-ARIA 1.0 autocomplete spec](https://github.com/department-of-veterans-affairs/va.gov-team/issues/1056#issuecomment-537096087) as Downshift v3 uses the WAI-ARIA 1.1 spec which creates challenges for screen readers.

Public Websites began implementation with [accessibility feedback from Jennifer](https://github.com/department-of-veterans-affairs/va.gov-team/issues/13343) which requested further testing with VoiceOver, TalkBack, NVDA, and JAWS.

This leads us to the [current spot check review for search discovery typeahead](https://github.com/department-of-veterans-affairs/va.gov-team/issues/16914#event-4070295479).

The existing implementation in review is not one of the two recommended implementations from [Crystabel's ticket: a datalist or downshift w/ WAI-ARIA 1.0 auto-complete](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/308).

@srsuddath clarified that the datalist option was not implemented because it handles most of its events under the hood, meaning theres no place to attach handlers for changes, clicks, selections, or any other events. For example:
- The preferred implementation will allow a user to click one of the suggestions from the drop down, and then automatically perform a search when that option is selected. With a datalist: when you select an option, it updates the input field, and there is no event to fire off the search on. This also erases the previous user input, which we prevents us from gathering useful analytics on how well typeahead is performing.
- With the current implementation (and hopefully the Downshift implementation as well) when an option is selected, we are first logging an event which captures all of the relevant analytics without destroying the user's input, and then we are firing off a search using the selected option as the query.

## Next Steps
~~1. We should review how the current implementation was created and why it did not follow existing recommendations from previous tickets as documented above.~~
2. If it is determined that special reasons existed for not following prior recommendations, we should explore those first and explore whether or not more accessible options exist. 👈 
~~3. If it is determined that no special reasons existed for not following prior recommendations, we should continue where @jenstrickland and @CrystabelReiter left off by conducting both usability and accessibility tests of the datalist and downshift implementations.~~

**Requirements for the new suggested solution**
- [x] Must be able to log an event that captures all of the relevant analytics without destroying the user's input.
- [x] Must be able to fire off a search using the selected option as the query.

## Exploring Solutions
[We've documented our research on accessible solutions in a google slide deck.](https://docs.google.com/presentation/d/1JSN9Bd7-4yOl8hVZS3zEhHuDxI3DMAYzZLCXYz9fMfs/edit?usp=sharing) Options explored include:
- datalists
- WAI-ARIA 1.0, 1.1, and 1.2 comboboxes
- downshift
- reach UI

## Final Recommendation
We recommend to upgrade Downshift to WAI-ARIA 1.2 Guidance for the following reasons:
- Similar work has been done before downgrading to 1.0 Guidance.
- Addresses concerns with both the WAI-ARIA 1.1 and 1.0 combobox patterns.
- Won’t add in a new library (e.g. Reach UI) that may add complexity or weight.
- Less issues on Safari w/VO compared to Reach UI and provides better instructions to users.

**Comment by srsuddath** [comment]
Hello @joshkimux and @jenstrickland I am catching up on this a bit. After reading through the above information, I am going to explore the downshift option. The datalist component does not support the event logging options that are needed for our event logging requirements, which is why this was implemented in a different way, additionally, it does not support many of the styling options so we have relatively little control as to its appearance. I was unaware of the previous discussions until very recently, so will have to explore this other option. Has the Downshift component been through accessibility testing?

**Comment by srsuddath** [comment]
Downshift component has been implemented, reviewed, and iterated over. ready to close this ticket

**Comment by srsuddath** [comment]
This issue has been resolved and we are clear to launch on Monday, see slack thread : [slack](https://dsva.slack.com/archives/C011K1VSTC3/p1608320422293600?thread_ts=1608308729.281300&cid=C011K1VSTC3)
